### PR TITLE
multiple commands, option parsing

### PR DESCRIPTION
**Pull request by mweitzel**
More useless subcommands.

Now distinguishes between options and args whether or not they are top level or nested command options.

### CODE REVIEW

File: cli/lib/app.rb [hunk 1]
```diff
@@ -1,26 +1,88 @@
...
+    cmd = cmds.first
+    cmd ||= "help"
+    raise "#{cmd} is not a subcommand" unless ["whoami", "help"].include?(cmd)
+    cmd
+  end
+
+  def config
+    @config ||= OpenStruct.new(
+      {
+        conf_location: "~/.cpci.conf"
```
Comment (jamandbees) [inline cli/lib/app.rb:76]: This is a total aside.

There's both linux standards and mac os standards for where those directories should be; I think mac os prefers ~/.Library/ and linux prefers a million things (Kali linux, for instance, asks you to pop 'em into ~/.config (https://kali.training/topic/the-filesystem/)) whereas tldp reckons that a dotfile is fine (https://tldp.org/LDP/Linux-Filesystem-Hierarchy/html/Linux-Filesystem-Hierarchy.html#home).

Who bloody knows what to use, but whatever it is we should have this string not be a constant.

File: cli/lib/app.rb [hunk 2]
```diff
@@ -34,15 +96,15 @@ class NotLoggedInException < StandardError
...
     found_name = creds["name"]
     raise NotLoggedInException.new unless found_name
     found_name
   end
 
   def creds
-    creds_path = path || "~/.cpci.conf"
+    raise "set path!" unless path
+    creds_path = path
     JSON.parse(File.read(creds_path))
```
Comment (jamandbees) [inline cli/lib/app.rb:108]: The only clue this file's intended to be in json format is right here where it's being parsed by a json formatter.

File: cli/lib/app.rb [hunk 3]
```diff
@@ -34,15 +96,15 @@ class NotLoggedInException < StandardError
 
 class Creds
```
Comment (jamandbees) [inline cli/lib/app.rb:97]: I think path be defined during class construction.

File: cli/lib/nested_parser.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
+require "optparse"
+require "ostruct"
+
+class OptParseX
```
Comment (jamandbees) [inline cli/lib/nested_parser.rb:4]: This class isn't used anywhere currently except for its tests; what is it? I think I got a clue from the file name; maybe it should be named `NestedParser`?

File: cli/test/test_nested_parser.rb [hunk 5]
```diff
@@ -0,0 +1,97 @@
+require "nested_parser"
+require "test_helper"
+
+describe "OptParseX" do
+  it "joins parsers" do
```
Comment (jamandbees) [inline cli/test/test_nested_parser.rb:5]: Why does it join parsers? What does joining parsers mean? The test reads like we're verifying that the option parser raises errors when there's an invalid option; it's pretty opaque.

File: cli/lib/nested_parser.rb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  attr_reader :ons
+  def initialize(&bloc)
```
Comment (jamandbees) [inline cli/lib/nested_parser.rb:6]: Naming a block 'bloc' isn't super intention revealing; is there a better name here?

File: cli/test/test_nested_parser.rb [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    p2 = OptParseX.new { |opts| opts.on("-bREQ") }
+    p3 = OptParseX.new { |opts| opts.on("-c") }
+
+    pp = p1.merge(p2, p3)
+    parsed = pp.parse("-a", "-b", "the_option", "the_arg", "-c", "-d")
+    assert_equal(["the_arg"], parsed.args)
+    assert_equal({:a => true, :b => "the_option", :c => true, :dry_run => true}, parsed.opts)
+  end
```
Comment (jamandbees) [inline cli/test/test_nested_parser.rb:96]: The names of all these tests blocks aren't descriptive, which is making it harder to understand why we're testing these things.

If I have something like:

```
def foo
  ["1", "2", "3"]
end
```
and a test like

```
it 'arrays' do
  foo === Array
end
```
the test isn't giving you any information other than what you can already see by reading the implementation; foo returns an array. 

For example, why is the `it "joins parsers"` test asserting that parsers are joined? What does that mean? The tests read like we're asserting that errors are raised when invalid options are handed in; could the name reflect that more?

The test `it "bloc runs"` has a name suggesting that we're verifying that the block functionality in ruby works when it's called. We already know that's true: it's been stable for a decade or two. What kinds of thing does it run? Why is it there? What's the error condition we're isolating, and the functionality that we're verifying?

File: cli/test/test_whoami.rb [hunk 8]
```diff
@@ -11,4 +11,22 @@
...
+    @app = FactoryBot.build(:app, creds: FactoryBot.build(:creds, name: "dingus"))
+    cmd = @app.run("whoami")
+    assert_includes(cmd.stdout, "dingus")
+  end
+
+  it "prints name" do
+    @app = FactoryBot.build(:app, creds: FactoryBot.build(:creds, name: "dingus"))
+    @app.run("whoami", "-t")
+    assert(@app.config.test_flag_set)
+  end
```
Comment (jamandbees) [inline cli/test/test_whoami.rb:25]: These two testa have the same name; what's the difference they're trying to validate?

File: cli/lib/app.rb (hunk 3 above)
Comment (mweitzel) [reply cli/lib/app.rb:97]: I'd like this too, it bothered me not to, but I wanted to get a better sense of the lifecycle of the app and it's components first.

Before it wasn't clear that I wanted to parse options, _then_ instantiate everything.

File: cli/lib/nested_parser.rb (hunk 4 above)
Comment (mweitzel) [reply cli/lib/nested_parser.rb]: I'm not sure the extent of the extra's well need so I was hesitant to name it "nested", I've refactored to extend the option parser instead of wrapping it. Will change the file name as well.

File: cli/test/test_nested_parser.rb (hunk 5 above)
Comment (mweitzel) [reply cli/test/test_nested_parser.rb]: renamed tests to illustrate intent

File: cli/lib/nested_parser.rb (hunk 6 above)
Comment (mweitzel) [reply cli/lib/nested_parser.rb:5]: `bloc` variable no longer needed with inheritance. super class doesn't name it

File: cli/test/test_nested_parser.rb (hunk 7 above)
Comment (mweitzel) [reply cli/test/test_nested_parser.rb]: Good pointy. I renamed and commented the first test. The following tests are largely edge-cases I wanted coverage for.

File: cli/test/test_whoami.rb (hunk 8 above)
Comment (mweitzel) [reply cli/test/test_whoami.rb]: fixed

File: cli/test/test_nested_parser.rb (hunk 7 above)
Comment (mweitzel) [reply cli/test/test_nested_parser.rb]: Also `it "bloc runs"` removed after refactor. It was redundant to test what is not a parent class (as opposed to a wrapped class)

**Review by jamandbees — APPROVED** [review]
Merging as it's more important to establish progress while we're prototyping.

Way to be tackling a new problem, @mweitzel !!
